lib/matchers/matchers.rb

```diff
@@ -338,3 +338,164 @@ def format_expectation(negate)
     end
   end
 end
+
+# The following are required for the PAM resource
+#
+# Copyright 2019 - Onyx Point, Inc.
+#
+# TODO explain somewhere that :all_with_args, :all_without_args, :all_with_integer_arg
+# will cause match_pam_rule to return true when there are no potential matches
+RSpec::Matchers.define :match_pam_rule do |expected|
+  def matching_integer_arg? (line)
+    line.module_arguments.any? do |arg|
+      key, value = arg.split('=')
+
+      value && (@args[:key] == key) && value.match?(/^-?\d+$/) &&
+        value.to_i.send(@args[:operator].to_sym, @args[:value])
+    end
+  end
+
+  match do |actual|
+    case @args_type
+    when :all_with_args, :all_without_args, :all_with_integer_arg
+      retval = true
+    when :any_with_args, :any_with_integer_arg
+      retval = false
+    end
+
+    if [:all_with_integer_arg, :any_with_integer_arg].include? @args_type
+      unless Numeric.method_defined?(@args[:operator])
+        fail("Error: Operator '#{@args[:operator]}' is an invalid numeric comparison operator.")
+      end
+    end
+
+    actual_munge = {}
+
+    @expected = expected.to_s
+
+    if @args_type
+      catch :stop_searching do
+        actual.services.each do |service|
+          expected_line = Pam::Rule.new(expected, {:service_name => service})
+
+          potentials = actual.find_all do |line|
+            line.match?(expected_line)
+          end
+
+          if potentials && !potentials.empty?
+            actual_munge[service] ||= []
+            actual_munge[service] += potentials.map(&:to_s)
+
+            potentials.each do |potential|
+              case @args_type
+              when :all_without_args
+                retval = !potential.module_arguments.join(' ').match?(@args)
+                throw :stop_searching unless retval
+              when :all_with_args
+                retval = potential.module_arguments.join(' ').match?(@args)
+                throw :stop_searching unless retval
+              when :all_with_integer_arg
+                retval = matching_integer_arg? potential
+                throw :stop_searching unless retval
+              when :any_with_integer_arg
+                retval = matching_integer_arg? potential
+                throw :stop_searching if retval
+              when :any_with_args
+                retval = potential.module_arguments.join(' ').match?(@args)
```

Comment:
Pretty sure I inherited the notation from the general matchers to keep it consistent. Honestly, some of the matcher bits were pulled together from examples because not much of it was super clear and relies on internal RSpec items that I didn't dig very far into.